app/javascript/components/users/SignInPage.spec.tsx

```diff
@@ -12,62 +15,154 @@ jest.mock('../../legacy_react/src/lib/api/sign_in');
+				fireEvent.click(button);
+			});
+
+			// was document.location.assign called with redirectUrl?
+			expect(locationAssignSpy).toHaveBeenCalledWith('redirectUrl');
+
+			const error = result.getByTestId('signInPageError');
+			expect(error).toBeEmptyDOMElement();
 		});
 
-		const error = result.getByTestId('signInPageError');
-		expect(error).toBeEmptyDOMElement();
 	});
 
 	it('signIn failed', async () => {
 		expect.hasAssertions();
-		const result = render(<Wrapper><SignInPage/></Wrapper>);
-
-		const button = result.getByTestId('signInButton');
-		// everytime you try to call the User SignIn API in this test, return a
-		// promise which resolves to {id: 1}
-		mockedWebUserSignIn.postSignIn.mockResolvedValue({id: 1});
-
-		// act puts all of the related React updates for the click event into a
-		// single update. Since fireEvent.click calls some promises, we need to make
-		// the callback a Promise and await on act. If we didn't, our test wouldn't
-		// wait for all the possible React changes to happen at once.
-		await act(async () => {
-			fireEvent.click(button);
+
+		locationAssign(async (locationAssignSpy:jest.SpyInstance<void, [url: string]>) => {
+
+			const result = render(<Wrapper><SignInPage redirectUrl={'redirectUrl'}/></Wrapper>);
+
+			const button = result.getByTestId('signInButton');
+			// everytime you try to call the User SignIn API in this test, return a
+			// promise which resolves to {id: 1}
+			mockedWebUserSignIn.postSignIn.mockResolvedValue({id: 1});
+
+			// act puts all of the related React updates for the click event into a
+			// single update. Since fireEvent.click calls some promises, we need to make
+			// the callback a Promise and await on act. If we didn't, our test wouldn't
+			// wait for all the possible React changes to happen at once.
+			await act(async () => {
+				fireEvent.click(button);
+			});
+
+			// was document.location.assign not called?
+			expect(locationAssignSpy).not.toHaveBeenCalledWith('redirectUrl');
+
+			const error = result.getByTestId('signInPageError');
+
+			// Sometimes because of React's rendering cycle, the changes haven't been
+			// made to the HTML by the time we need to expect. waitFor tries some sort
+			// function multiple times until it passes (or 5 seconds have passed)
+			waitFor(() => expect(error).toHaveTextContent("Ermahgerd! We had an error!"));
 		});
+	});
+
+	it('renders signInComponent Correctly', () => {
+		const { getByTestId } = render(<Wrapper><SignInPage redirectUrl={'redirectUrl'}/></Wrapper>);
+		expect(getByTestId("SignInComponent")).toBeTruthy;
+	});
+
+});
+
 
-		const error = result.getByTestId('signInPageError');
+describe('Links', () => {
+	it('renders forgot password Link', () => {
+		const { getByTestId } = render(<Wrapper><SignInPage redirectUrl={'redirectUrl'}/></Wrapper>);
+		fireEvent.click(getByTestId("passwordTest"));
+		getByTestId('passwordTest').click();
+		expect(getByTestId('passwordTest')).toBeInTheDocument();
```

Comment:
So, I want you to step back and think about what we'd want to test about a forgot password link. Don't immediately worry about the HOW, worry about what you could test about it.

* Does the link exist?
* Does it have the correct text?
* Does it go to the correct url when clicked?

This test only verifies that the first one is true. Additionally, remember that using getByTestId is a last resort if we can't get it another way. We want to try to test it as closely to how a user would test it. They would look for the text `Forgot Password?` So can we use `getByText('Forgot Password?')` to get the link element? If not, why?

Then let's check that it actually goes to the url that we expect it to. That's what `locationAssign`, as described above is used for.